.github/workflows/build_develop.yml

```diff
@@ -34,28 +38,81 @@ jobs:
           SENTRY_URL: ${{ secrets.SENTRY_URL }}
           SENTRY_ORG: element
           SENTRY_PROJECT: riot-web
+          # We have to update the CSP until Cloudflare Pages supports proxying redirects
+          CSP_EXTRA_SOURCE: ${{ env.R2_PUBLIC_URL }}
 
       - run: mv dist/element-*.tar.gz dist/develop.tar.gz
 
-      # We keep the latest develop.tar.gz as the artifact uploaded later expires after 24 and requires auth to download
-      # Element Desktop's fetch script uses this tarball to fetch latest develop to build Nightlies.
-      - name: Deploy develop.tar.gz to Github Pages
-        uses: JamesIves/github-pages-deploy-action@v4
-        with:
-          folder: dist
-          single-commit: true
+      # TODO: remove this once the element-desktop fetch script is updated to grab develop.element.io/develop.tar.gz
+#      - name: Deploy develop.tar.gz to Github Pages
+#        uses: JamesIves/github-pages-deploy-action@v4
+#        with:
+#          folder: dist
+#          single-commit: true
 
       - uses: actions/upload-artifact@v3
         with:
           name: webapp
           path: dist/develop.tar.gz
           retention-days: 1
+          
+      - name: Extract webapp
+        run: |
+          mkdir _deploy
+          tar xf dist/develop.tar.gz -C _deploy --strip-components=1
+          
+      - name: Copy config
+        run: cp element.io/develop/config.json _deploy/config.json
+
+      - name: Populate 404.html
+        run: echo "404 Not Found" > _deploy/404.html
+
+      - name: Populate _headers
+        run: cp .github/cfp_headers _deploy/_headers
+
+      - name: Populate _redirects
+        run: |
+          echo "/develop.tar.gz $R2_PUBLIC_URL/develop.tar.gz 301" >> _deploy/_redirects
+          for bundle in $(aws s3 ls s3://$R2_BUCKET/bundles/ --endpoint-url $R2_URL --region=auto | awk '{print $2}'); do
+            echo "/bundles/${bundle}* $R2_PUBLIC_URL/bundles/${bundle}:splat 301" >> _deploy/_redirects
+          done
+          cat _deploy/_redirects
+        env:
```

Comment:
Why? No one should have the R2 and CFP creds to be able to run this by locally

Reply:
>  having to read many scripts side-by-side to debug a workflow is a nightmare imo

personally what I find a nightmare is trying to read stuff that ought to be factored out but is arbitrarily scattered into distinct steps in a massive yaml file.

You might as well ask "why should I factor functionality out to separate functions if it's only used in one place?". It makes it *easier* to read by collecting related bits of functionality together, and encapsulating them into a coherent interface. Then, when I come to debug problems with this workflow, I don't need to worry about the `build_cfp_deployment` script unless there's something specifically wrong with the shape of that deployment.

Another thing: there are 8 shell commands here. Why are they divided into 5 GHA steps, rather than 1 or 8? It feels pretty arbitrary.

I'll not insist, but it would certainly be my recommendation to factor it out to a separate script.

Reply:
personally, I'd rather not have to deal with a context switch when debugging CI: opening a thousand editors to figure out what the thing even does isn't great.

Reply:
if you have to open even one (more) editor to figure out what it does, never mind 1000, you're doing it wrong.

But it feels like I'm in the minority here, so I guess I'll get back in my box.